File: android/src/main/java/expo/modules/brownfield/ExpoBrownfieldModule.kt
package expo.modules.brownfield

import androidx.activity.ComponentActivity
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

object BrownfieldNavigationState {
  var nativeBackEnabled = true
}

class ExpoBrownfieldModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoBrownfieldModule")

    Function("popToNative") { animated: Boolean ->
      appContext.currentActivity?.runOnUiThread {
        val componentActivity = appContext.currentActivity as? ComponentActivity
        if (componentActivity != null) {
          val enabled = BrownfieldNavigationState.nativeBackEnabled
          BrownfieldNavigationState.nativeBackEnabled = true
          componentActivity.onBackPressedDispatcher?.onBackPressed()
          BrownfieldNavigationState.nativeBackEnabled = enabled
        }
      }
    }

    Function("setNativeBackEnabled") { enabled: Boolean ->
      BrownfieldNavigationState.nativeBackEnabled = enabled
    }
  }
}

File: ios/ExpoBrownfieldModule.swift
import ExpoModulesCore

public class ExpoBrownfieldModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoBrownfieldModule")

    Function("popToNative") { (animated: Bool) in
      DispatchQueue.main.async {
        NotificationCenter.default.post(
          name: Notification.Name("popToNative"),
          object: nil,
          userInfo: ["animated": animated]
        )
      }
    }

    Function("setNativeBackEnabled") { (enabled: Bool) in
      DispatchQueue.main.async {
        NotificationCenter.default.post(
          name: Notification.Name("setNativeBackEnabled"),
          object: nil,
          userInfo: ["enabled": enabled]
        )
      }
    }
  }
}
